<launch>
    <arg name="signaling_server_hostname" default="localhost"/>
    <arg name="signaling_server_port" default="8080"/>
    <arg name="signaling_server_password" default="abc"/>
    <arg name="is_stand_alone" default="false"/>

    <arg name="output" default="log"/>

    <!-- Signaling server (if stand_alone)-->
    <include if="$(arg is_stand_alone)" file="$(find opentera_webrtc_demos)/launch/opentera_signaling_server.launch">
        <arg name="signaling_server_port" value="$(arg signaling_server_port)" />
        <arg name="signaling_server_password" value="$(arg signaling_server_password)" />
        <arg name="signaling_server_robot_type" value="TTOP" />
        <arg name="python_executable" value="python3.8" />
        <arg name="use_tls" value="true" />
        <arg name="certificate" value="~/.ros/opentera/certs/server-cert.pem" />
        <arg name="key" value="~/.ros/opentera/certs/server-key.pem" />
        <arg name="output" value="$(arg output)" />
    </include>

    <!-- OpenTera Client (if not stand_alone)-->
    <include unless="$(arg is_stand_alone)" file="$(find opentera_client_ros)/launch/client.launch">
        <arg name="opentera_client_config_file" value="~/.ros/opentera/client_config.json"/>
        <arg name="output" value="$(arg output)"/>
    </include>

    <!-- Front camera -->
    <include file="$(find opentera_webrtc_ros)/launch/ros_stream_client.launch">
        <arg name="name" value="topic_streamer1"/>
        <arg name="output" value="$(arg output)"/>
        <arg name="is_stand_alone" value="$(arg is_stand_alone)"/>
        <arg name="can_send_audio_stream" value="true"/>
        <arg name="can_receive_audio_stream" value="true"/>
        <arg name="can_send_video_stream" value="true"/>
        <arg name="can_receive_video_stream" value="true"/>
        <arg name="is_screen_cast" value="false"/>
        <arg name="needs_denoising" value="false"/>
        <arg name="server_url" value="https://localhost:$(arg signaling_server_port)"/>
        <arg name="client_name" value="Robot Camera"/>
        <arg name="room_name" value="VideoConf"/>
        <arg name="room_password" value="$(arg signaling_server_password)"/>
        <arg name="input_camera_topic" value="/camera/color/image_raw"/>
        <arg name="input_audio_topic" value="/audio_webrtc_outgoing_signed_16_48000"/>
        <arg name="output_audio_mixed_topic" value="/audio_webrtc_incoming_signed_16_48000"/>
        <arg name="verify_ssl" value="$(eval not arg('is_stand_alone'))"/>
    </include>

    <!-- Map image stream -->
    <include file="$(find opentera_webrtc_ros)/launch/ros_stream_client.launch">
        <arg name="name" value="topic_streamer2"/>
        <arg name="output" value="$(arg output)"/>
        <arg name="is_stand_alone" value="$(arg is_stand_alone)"/>
        <arg name="can_send_audio_stream" value="false"/>
        <arg name="can_receive_audio_stream" value="false"/>
        <arg name="can_send_video_stream" value="true"/>
        <arg name="can_receive_video_stream" value="false"/>
        <arg name="is_screen_cast" value="true"/>
        <arg name="needs_denoising" value="false"/>
        <arg name="server_url" value="https://localhost:$(arg signaling_server_port)"/>
        <arg name="client_name" value="Map image"/>
        <arg name="room_name" value="Map"/>
        <arg name="room_password" value="$(arg signaling_server_password)"/>
        <arg name="input_camera_topic" value="/map_image_generator/image"/>
        <arg name="verify_ssl" value="$(eval not arg('is_stand_alone'))"/>
    </include>

    <!-- Data Channel  -->
    <include file="$(find opentera_webrtc_ros)/launch/ros_data_channel_client.launch">
        <arg name="name" value="data_streamer"/>
        <arg name="output" value="$(arg output)"/>
        <arg name="is_stand_alone" value="$(arg is_stand_alone)"/>

        <arg name="server_url" value="https://localhost:$(arg signaling_server_port)"/>
        <arg name="client_name" value="Robot Data Channel"/>
        <arg name="room_name" value="Teleop"/>
        <arg name="room_password" value="$(arg signaling_server_password)"/>
        <arg name="verify_ssl" value="$(eval not arg('is_stand_alone'))"/>
    </include>

    <!-- Json Message Handler -->
    <include file="$(find opentera_webrtc_ros)/launch/ros_json_data_handler.launch">
        <arg name="name" value="json_data_handler"/>
        <arg name="output" value="$(arg output)"/>
        <arg name="input_topic" value="/webrtc_data_incoming"/>
        <arg name="linear_multiplier" value="0.5"/>
        <arg name="angular_multiplier" value="0.35"/>
        <arg name="cmd_vel_topic" value="/teleoperation/cmd_vel"/>
    </include>

    <!-- Playback for audio coming from webrtc -->
    <node pkg="audio_utils" type="alsa_playback_node" name="opentera_playback_node" respawn="true">
        <param name="device" value="default"/>
        <param name="format" value="signed_16"/>
        <param name="channel_count" value="1"/>
        <param name="sampling_frequency" value="48000"/>
        <param name="frame_sample_count" value="480"/>
        <param name="latency_us" value="100000"/>

        <remap from="audio_in" to="audio_webrtc_incoming_signed_16_48000"/>
    </node>

    <!-- Audio resampling for audio going to webrtc -->
    <node pkg="audio_utils" type="resampling_node.py" name="resampling_node">
        <param name="input_format" value="signed_16"/>
        <param name="output_format" value="signed_16"/>
        <param name="channel_count" value="1"/>
        <param name="input_sampling_frequency" value="44100"/>
        <param name="output_sampling_frequency" value="48000"/>
        <param name="input_frame_sample_count" value="2048"/>

        <remap from="audio_in" to="audio_signed_16_44100"/>
        <remap from="audio_out" to="audio_webrtc_outgoing_signed_16_48000"/>
    </node>

    <!-- Robot status -->
    <node name="robot_status" type="robot_status.py" pkg="t_top" />

    <!-- OpenTera Robot GUI -->
    <node pkg="opentera_webrtc_robot_gui" type="opentera_webrtc_robot_gui_node" name="robot_gui" output="$(arg output)" respawn="true">
        <remap from="/front_camera/image_raw" to="/camera/color/image_raw"/>
    </node>

</launch>
